File: app_native_backup/src/main/java/com/example/photoorganizer/data/worker/TrashCleanupWorker.kt
package com.example.photoorganizer.data.worker

import android.content.Context
import android.util.Log
import androidx.hilt.work.HiltWorker
import androidx.work.CoroutineWorker
import androidx.work.Data
import androidx.work.PeriodicWorkRequestBuilder
import androidx.work.WorkRequest
import androidx.work.WorkerParameters
import com.example.photoorganizer.data.local.safe.TrashManager
import dagger.assisted.Assisted
import dagger.assisted.AssistedInject
import java.util.concurrent.TimeUnit

/**
 * Daily worker for automatic trash cleanup.
 *
 * Runs periodically (every 24 hours) to permanently delete photos
 * that have exceeded the 7-day retention period. Uses WorkManager
 * for reliable background execution with battery-aware constraints.
 *
 * This worker is scheduled after onboarding completion and persists
 * across app restarts using unique work name "trash_cleanup".
 *
 * @see TrashManager
 * @see TrashItem
 */
@HiltWorker
class TrashCleanupWorker @AssistedInject constructor(
    @Assisted context: Context,
    @Assisted params: WorkerParameters,
    private val trashManager: TrashManager
) : CoroutineWorker(context, params) {

    companion object {
        private const val TAG = "TrashCleanupWorker"

        /** Unique work name for trash cleanup scheduling */
        const val WORK_NAME = "trash_cleanup"

        /** Repeat interval in hours (24 hours = daily) */
        const val REPEAT_INTERVAL_HOURS = 24L

        /** Initial delay in minutes (start 1 hour after app launch/onboarding) */
        const val INITIAL_DELAY_MINUTES = 60L

        /** Key for deleted count in output data */
        const val OUTPUT_DELETED_COUNT = "deleted_count"

        /**
         * Create a periodic work request for daily trash cleanup.
         *
         * Uses battery-aware constraints to avoid running during low battery.
         * Existing work with the same name is kept (not replaced).
         */
        fun createWorkRequest(): WorkRequest {
            return PeriodicWorkRequestBuilder<TrashCleanupWorker>(
                REPEAT_INTERVAL_HOURS,
                TimeUnit.HOURS
            )
                .setInitialDelay(INITIAL_DELAY_MINUTES, TimeUnit.MINUTES)
                .addTag(TAG)
                .build()
        }
    }

    /**
     * Execute trash cleanup.
     *
     * Gets expired items from TrashManager and permanently deletes them.
     * Returns success with deleted count regardless of individual item
     * failures - each item failure is logged but doesn't block others.
     *
     * @return Result.success with deleted count, or Result.retry if critical error
     */
    override suspend fun doWork(): Result {
        Log.d(TAG, "Starting daily trash cleanup")

        return try {
            // Get current timestamp
            val now = System.currentTimeMillis()

            // Get expired items (older than 7 days)
            val expiredItems = trashManager.getExpiredItems(now)
            Log.d(TAG, "Found ${expiredItems.size} expired items to delete")

            if (expiredItems.isEmpty()) {
                Log.d(TAG, "No expired items to clean up")
                return Result.success(
                    Data.Builder()
                        .putInt(OUTPUT_DELETED_COUNT, 0)
                        .build()
                )
            }

            // Delete each expired item
            var deletedCount = 0
            var failedCount = 0

            for (item in expiredItems) {
                trashManager.permanentlyDelete(item).fold(
                    onSuccess = {
                        deletedCount++
                        Log.d(TAG, "Deleted: ${item.fileName}")
                    },
                    onFailure = { e ->
                        failedCount++
                        Log.e(TAG, "Failed to delete: ${item.fileName}", e)
                        // Continue with other items - don't let one failure block cleanup
                    }
                )
            }

            Log.d(TAG, "Cleanup complete: $deletedCount deleted, $failedCount failed")

            // Return success even if some items failed - they'll be retried next run
            val outputData = Data.Builder()
                .putInt(OUTPUT_DELETED_COUNT, deletedCount)
                .putInt("failed_count", failedCount)
                .build()

            Result.success(outputData)

        } catch (e: Exception) {
            Log.e(TAG, "Critical error during trash cleanup", e)
            // Retry on critical errors (database connectivity, etc.)
            Result.retry()
        }
    }
}

File: lib/data/workers/trash_cleanup_worker.dart
import 'package:flutter/foundation.dart';
import 'package:workmanager/workmanager.dart';
import '../database/database_service.dart';
import '../repositories/trash_repository.dart';
import '../platform/file_operation_service.dart';
import '../services/trash_service.dart';
import '../../domain/models/trash_item.dart';

/// Background worker for daily trash cleanup
/// 
/// Runs once per day to permanently delete trash items that have expired
/// (older than 7 days). This implements the automatic cleanup requirement
/// for the trash system.
/// 
/// The worker is scheduled after onboarding completion and uses WorkManager's
/// periodic task scheduling with KEEP policy to prevent duplicate scheduling.
class TrashCleanupWorker {
  static const String taskName = 'trashCleanup';

  /// Callback dispatcher for WorkManager background execution
  /// 
  /// This must be a top-level function annotated with @pragma('vm:entry-point')
  /// to ensure it's not tree-shaken during release builds.
  @pragma('vm:entry-point')
  static void callbackDispatcher() {
    Workmanager().executeTask((task, inputData) async {
      if (task == taskName) {
        debugPrint('TrashCleanupWorker: Starting cleanup task');
        
        try {
          // Initialize services
          final db = DatabaseService();
          final fileService = FileOperationService();
          final trashRepo = TrashRepository(db);
          final trashService = TrashService(
            fileOperationService: fileService,
            trashRepository: trashRepo,
          );
          
          // Get expired items
          final expiredItems = await trashRepo.getExpiredItems();
          debugPrint('TrashCleanupWorker: Found ${expiredItems.length} expired items');
          
          // Delete each expired item permanently
          int deletedCount = 0;
          for (final item in expiredItems) {
            try {
              final result = await trashService.permanentlyDelete(item);
              if (result.isSuccess) {
                deletedCount++;
                debugPrint('TrashCleanupWorker: Deleted ${item.fileName}');
              } else {
                debugPrint('TrashCleanupWorker: Failed to delete ${item.fileName}: ${result.error}');
              }
            } catch (e) {
              debugPrint('TrashCleanupWorker: Error deleting ${item.fileName}: $e');
            }
          }
          
          debugPrint('Trash cleanup: deleted $deletedCount items');
          return true;
        } catch (e) {
          debugPrint('TrashCleanupWorker: Error during cleanup: $e');
          return false;
        }
      }
      
      return true;
    });
  }

  /// Schedule daily trash cleanup
  /// 
  /// Uses WorkManager's periodic task with 1-day frequency.
  /// The KEEP policy prevents duplicate scheduling if called multiple times.
  /// 
  /// Requires battery not low constraint to avoid draining user's device.
  static Future<void> scheduleDailyCleanup() async {
    debugPrint('TrashCleanupWorker: Scheduling daily cleanup');
    
    await Workmanager().registerPeriodicTask(
      taskName,
      taskName,
      frequency: const Duration(days: 1),
      constraints: Constraints(
        requiresBatteryNotLow: true,
        requiresStorageNotLow: false, // Can run even if storage is low (cleanup frees space)
      ),
      existingWorkPolicy: ExistingWorkPolicy.keep, // Prevent duplicates
      initialDelay: const Duration(hours: 1), // Wait 1 hour after app start
    );
    
    debugPrint('TrashCleanupWorker: Daily cleanup scheduled');
  }

  /// Cancel the scheduled cleanup worker
  static Future<void> cancelCleanup() async {
    debugPrint('TrashCleanupWorker: Cancelling scheduled cleanup');
    await Workmanager().cancelByUniqueName(taskName);
  }

  /// Manually trigger cleanup (for testing or user-initiated cleanup)
  static Future<int> runCleanupNow() async {
    debugPrint('TrashCleanupWorker: Running manual cleanup');
    
    try {
      final db = DatabaseService();
      final fileService = FileOperationService();
      final trashRepo = TrashRepository(db);
      final trashService = TrashService(
        fileOperationService: fileService,
        trashRepository: trashRepo,
      );
      
      final expiredItems = await trashRepo.getExpiredItems();
      int deletedCount = 0;
      
      for (final item in expiredItems) {
        final result = await trashService.permanentlyDelete(item);
        if (result.isSuccess) {
          deletedCount++;
        }
      }
      
      debugPrint('TrashCleanupWorker: Manual cleanup deleted $deletedCount items');
      return deletedCount;
    } catch (e) {
      debugPrint('TrashCleanupWorker: Manual cleanup failed: $e');
      return 0;
    }
  }
}
